### ISSUE TITLE
Support non-null/non-empty constraints in the data model

### ISSUE THREAD

**Issue by JoelRichardson** [intermine/intermine#1092]
A simple proposal for allowing models to specify that specific fields (attributes, references, or collections) are not null. There are three parts to this:

1. In model(-additions) files:  
 * Support "allowNulls" boolean attribute for attribute, reference, and collections tags. Default is "true", which is backwards compatible with current behavior. Existing models are completely unaffected. Can set to "false" for any attribute, reference, or collection. (Detail: for collections, maybe have a different attribute, "allowEmpty"? Or just use "allowNulls", but understand it to mean empty/nonempty.)
1. At build time:
 * New postprocess to verify. Build fails if any non-null/non-empty constraint is violated.
1. At runtime:
 * Generated queries use INNER joins when reference/collection has non-null/non-empty constraint. Prime example: the queries that are generated for displaying collections on report pages. (This is the real reason for the proposal.)

**Backstory**: I have complained for years about the way InterMine handles outer joins in the web interface. Why:
* You cannot sort or filter on a column that is reached through an outer join. 
	This severely hampers usability. In fact, filtering is just plain weird. It doesn't 
	actually filter anything; you still get all the rows back, but the filtered column now has lots of nulls
	(the rows that should have gone away). And since you can't sort on that column, you can't find the 
	non-filtered rows except by paging through all the results.
* There are many, many cases where I **know** a reference is not null. 
	InterMine's insistence on using outer joins is overconservative. And since the display of these
	columns is (in my book) broken, the default display of collections is often unusable. It forces us to
	hide these displays and write templates to take their place.
   
A previous issue (#441) proposed allowing the user to specify in webconfig-model.xml
   that the join should be inner. But there ended up being questions about how multiple
   fieldconfig's would or wouldn't play nicely together. The thing just petered out.

Well, I just got bit again. Without going into details, some new data is being loaded and we were just going to use default collection displays on report pages. In testing it out, I was reminded of this inner/outer issue.  In rethinking the issue and reviewing #441, it occurred to me that InterMine uses outer joins based on the conservative assumption that references can be null, and we want to avoid data drop-out. But what if InterMine knew that a reference can't be null? Couldn't it then use an inner join?

**Comment by julie-sullivan** [comment]
A good idea! 

What would be even better is if it didn't require configuration, e.g. the query system automatically knew if there were values or not. If there were no nulls, inner join. This sounds expensive to calculate but it might not be. We'd only have to run it once and cache the results.  

@joshkh How difficult would this be to check a relationship before running a query? This is for collections on the report page.

@JoelRichardson Would defaulting to be inner joins be better for you? What if we implement #441 and allow folks to configure joins to be OUTER if necessary. Would that solve your problem?

**Comment by JoelRichardson** [comment]
Hi Julie,
Thanks! I actually think this belongs in the model rather than being automatically inferred. Being in the model gives the dev an important measure of control that is verifiable and predictable. As for defaulting to inner joins in webconfig-model, that might solve my problem, but would change the behavior for everyone, which should be avoided I think. Also, allowing an outer attribute would lead to all the same questions that caused #441 to fizzle. Does that make sense? Hope all is well! Cheers, Joel

Sent from my iPhone

On Aug 5, 2015, at 5:30 AM, Julie Sullivan <notifications@github.com<mailto:notifications@github.com>> wrote:


A good idea!

What would be even better is if it didn't require configuration, e.g. the query system automatically knew if there were values or not. If there were no nulls, inner join. This sounds expensive to calculate but it might not be. We'd only have to run it once and cache the results.

@joshkh<https://github.com/joshkh> How difficult would this be to check a relationship before running a query? This is for collections on the report page.

@JoelRichardson<https://github.com/JoelRichardson> Would defaulting to be inner joins be better for you? What if we implement #441<https://github.com/intermine/intermine/issues/441> and allow folks to configure joins to be OUTER if necessary. Would that solve your problem?

-
Reply to this email directly or view it on GitHub<https://github.com/intermine/intermine/issues/1092#issuecomment-127932850>.

The information in this email, including attachments, may be confidential and is intended solely for the addressee(s). If you believe you received this email by mistake, please notify the sender by return email as soon as possible.

**Comment by julie-sullivan** [comment]
Is this a duplicate of  #441?

**Comment by JoelRichardson** [comment]
No, it is a more general solution that renders #441 unnecessary.

**Comment by julie-sullivan** [comment]
Moving to 2.0. Will be done in concert with #1148 - updating how the integration keys work.
